### ISSUE TITLE
Strange characters on refresh - Font Awesome 5 Pro

### ISSUE THREAD

**Issue by templeman15** [FortAwesome/Font-Awesome#17644]
**Describe the bug**
On random page loads or refreshing of page strange characters are displayed instead of the font awesome icons. Seems to very closely related to #15010 & #17611. However, it happens when using both `fa` classes and setting the content manually in css. 

There is no trace of FA4 in the project. 

**To Reproduce**

SCSS:
<img width="337" alt="Screen Shot 2021-02-17 at 1 25 31 PM" src="https://user-images.githubusercontent.com/3027970/108270106-d23ded80-7123-11eb-82c8-87e9cf43f8c0.png">

Expected:
<img width="168" alt="Screen Shot 2021-02-17 at 1 26 01 PM" src="https://user-images.githubusercontent.com/3027970/108270143-e2ee6380-7123-11eb-9daa-14721387611a.png">

Actual:
<img width="176" alt="Screen Shot 2021-02-17 at 1 25 53 PM" src="https://user-images.githubusercontent.com/3027970/108270163-ed106200-7123-11eb-985c-bad735f8fc2c.png">

**Expected behavior**
Expected behavior is that the font awesome icons would display correctly.

**Version and implementation**
Version: Font Awesome Pro 5.14.0
Browser and version: Google Chrome 88.0.4324.150
OS: Mac OS 10.15.7

- [ ] SVG with JS
- [x] Web Fonts with CSS
- [ ] SVG Sprites
- [ ] On the Desktop

**Bug report checklist**
- [x] I have filled out as much of the above information as I can
- [ ] I have included a test case because my odds go _way_ up that the team can fix this when I do
- [ ] I have [searched for existing issues](https://github.com/FortAwesome/Font-Awesome/issues) and to the best of my knowledge this is not a duplicate

**Comment by tagliala** [comment]
Hi!

Thanks for being part of the Font Awesome Community.

Are you using self-hosted icons?

**Comment by templeman15** [comment]
> Hi!
> 
> Thanks for being part of the Font Awesome Community.
> 
> Are you using self-hosted icons?

@tagliala yes we are! Thanks for the quick response :)

**Comment by tagliala** [comment]
Could you please try with the Pro CDN (not the kit, just the link to the plain css) and attempt to reproduce?

https://fontawesome.com/account/cdn

**Comment by tagliala** [comment]
Hi @templeman15

> SCSS:

I just read and realised that it was a screenshot of the `SCSS`, and that doesn't help. Any chance to get a look at the CSS output served to the browser by your application?

According to the Sass library used by your application, you may have wrong CSS results:

![image](https://user-images.githubusercontent.com/556268/108423987-db6cae80-7238-11eb-850f-633fe773c781.png)

**Comment by templeman15** [comment]
Hello @tagliala!

Here is the output served from the app (this is when the page loads correctly):
<img width="658" alt="Screen Shot 2021-02-18 at 1 49 52 PM" src="https://user-images.githubusercontent.com/3027970/108427571-3929d800-71f2-11eb-926d-0ef4076cd3b9.png">

Here is the output css from the browser (this is when the strange character occurs):

<img width="451" alt="Screen Shot 2021-02-18 at 2 02 43 PM" src="https://user-images.githubusercontent.com/3027970/108427474-126ba180-71f2-11eb-9d68-64514669d8a6.png">

**Comment by tagliala** [comment]
Ok, thanks.

Please switch to a different Sass implementation, make sure that your CSS output contains properly encoded characters

Meanwhile, closing here

**Comment by dkniffin** [comment]
@tagliala I am also seeing this behavior intermittently. 

Can you clarify: is there an issue using fontawesome with dart-sass?

**Comment by dkniffin** [comment]
After some more investigation, this might be due to us setting `$fa-font-display: swap`, as touched on in #14387

**Comment by tagliala** [comment]
> Can you clarify: is there an issue using fontawesome with dart-sass?

Not with Font Awesome. There are some issues with escaping with some versions of Dart Sass, please take another look at https://github.com/FortAwesome/Font-Awesome/issues/17644#issuecomment-781647877

There is no Font Awesome involved in this simple snippet:

```scss
.search-icon {
    &::before {
        content: "\f002";
    }
}
```

If the content translates to `content: "";`, then you may experience issues and I advise to double check the output. It should look like this: 

https://github.com/FortAwesome/Font-Awesome/blob/fcec2d1b01ff069ac10500ac42e4478d20d21f4c/css/fontawesome.css#L3396-L3397

> After some more investigation, this might be due to us setting $fa-font-display: swap, as touched on in #14387

You should not use that value for Font Icons: GoogleChrome/lighthouse#10127

**Comment by tagliala** [comment]
Hi @dkniffin 

I've investigated and

> Can you clarify: is there an issue using fontawesome with dart-sass?

Yes, there might be, in the case that a Font Awesome SCSS compiled to CSS with Dart Sass is not correctly interpreted as UTF-8 by the browser.

Although this may be considered a browser issue, I think that we should do our best to prevent it and provide ASCII output.

This will be hopefully fixed (again) in FA v6. I'm going to reopen this issue for tracking purposes

Since this will be a breaking change in v5, so I don't know what the team will decide.

The workaround is to  override all the `$fa-var-*` variables in `_variables.scss` and the `fa-content` mixin from `_mixins.scss`

Ref:
- https://github.com/sass/dart-sass/issues/568#issuecomment-770105804
- https://github.com/sass/dart-sass/issues/568#issuecomment-562086079

**Comment by tagliala** [comment]
The breaking change happened between dart-sass 1.10.4 and 1.11.0, but I can't bisect to the specific change because I do not have a proper dart build system

**Comment by tagliala** [comment]
### Solution

if you have PUA codes (e.g., `content: ""` instead of `content: "\f26e"`) in the output CSS, check that:
1. The served CSS file starts with `@charset "UTF-8";`; **AND**
2. The HTML file referring the CSS has the `<meta charset="utf-8">` meta tag

https://github.com/FortAwesome/Font-Awesome/issues/14660#issuecomment-790475966

# ⚠️  update

Unfortunately this is not enough. It may fix 99% of the cases, but not 100%

**Comment by pydolan** [comment]
FWIW, I still see the odd characters with FA 5.14.0 on a site with `<meta charset="utf-8">` in the header. I haven't tried adding `@charset` to the CSS file, but that's on my TODO list. (I'm seeing it in Chrome 90.x on Mac Big Sur)

With disclaimer about my limited browser internals knowledge, could it be a timing issue, where the font svg/wolff/ttf file isn't loading in time before the HTML renders the page?

**Comment by tagliala** [comment]
I don't know and I don't think so. `<meta charset="utf-8">` should be the first tag in `<head>`

**Comment by Arecsu** [comment]
I had this problem for so long and apparently I got rid of it by adding `@charset "UTF-8";` at the beginning of my CSS file as suggested above. I already had `<meta charset="utf-8">` in the html head, but seems like it wasn't enough.

Tested it plenty of times with different scenarios and I don't get the weird characters anymore. Everything works great.

**Comment by tagliala** [comment]
> I had this problem for so long and apparently I got rid of it by adding @charset "UTF-8"; at the beginning of my CSS file as suggested above. I already had <meta charset="utf-8"> in the html head, but seems like it wasn't enough.

Thanks for this heads-up

**Comment by mayankchd** [comment]
We had this problem too, investigated a bit and then figured out that this problem started happening after we switched to `dart-scss`. Looks like `dart-scss` by default in prod mode removes the `@chatset`.  This [issue](https://github.com/webpack-contrib/sass-loader/issues/973) in sass-loader has a [solution](https://github.com/webpack-contrib/sass-loader/issues/973#issuecomment-871402666)

**Comment by tagliala** [comment]
Unfortunately this did not meet the 6.x milestone because of backward compatibility.

This should only affect some users using both SCSS and dart-sass. 

I suggest to double check that both HTML and CSS (HTML is not enough) files have the correct charset. Some build system like webpack may strip out the `@charset` from the compiled css. You may want to use another plugin to write it back in the file

**Comment by danon** [comment]
> [...] This [issue](https://github.com/webpack-contrib/sass-loader/issues/973) in sass-loader has a [solution](https://github.com/webpack-contrib/sass-loader/issues/973#issuecomment-871402666)

More of an workaround, since it suggests disabling compression. Hardly a solution.

**Comment by tagliala** [comment]
This should have been fixed in `dart-sass`'s 1.38.0 release

https://github.com/sass/dart-sass/releases/tag/1.38.0

sass/dart-sass#1430

dart-sass users, please give it a try

I'm personally not happy with the result, I would like to see this "fixed" in FA, so I'm leaving this open for the moment

**Comment by danon** [comment]
@tagliala We're using `dart-sass` version `1.49` and the bug still happens sometimes.

**Comment by tagliala** [comment]
@Danon could you please provide an excerpt from the output CSS file where icons are defined?

I'm interested in knowing if the output is using unicode characters (`""`) or their escaped form (`"\f26e"`)

**Comment by danon** [comment]
> @Danon could you please provide an excerpt from the output CSS file where icons are defined?
> 
> I'm interested in knowing if the output is using unicode characters (`""`) or their escaped form (`"\f26e"`)

Can't since it happens randomly. About once a week for me.

**Comment by tagliala** [comment]
@Danon sorry, maybe I wasn't clear

I've asked an excerpt of the CSS file containing the rule of a Font Awesome icon taken from the website where you are experiencing the bug

In other words, the following code:

https://github.com/FortAwesome/Font-Awesome/blob/307f2c15e2d3bdf465b43df0d1486dcd249df2a1/css/fontawesome.css#L547-L548

but taken from your CSS

**Comment by ghost** [comment]
@tagliala I'm using Font Awesome with [dart-sass](https://github.com/sass/dart-sass) to compile the SCSS files. And of course, I'm experiencing the same issue, randomly. For some reason, the issue happens more often when the web console is open or being opened.

Here is the code you asked @Danon for, but from my CSS output:

```css
.fa-anchor::before {
    content: "";
}
```

Screenshot:
![bug](https://user-images.githubusercontent.com/59090546/160262402-3aa4af44-1f78-42e3-8b6d-1934581f1052.png)

Hope this can help.

**Comment by tagliala** [comment]
Yes, thanks.

We are trying to sort this out without breaking backward compatibility but it is a little bit trickier

As per dart sass, I can see that this has been partially fixed in 1.38.0, please make sure that you are using >= 1.38.0

There is also a workaround, but we don't know yet if it will fix 100% of the cases and requires effort on developer side. If you want to test, I've provided a step-by-step guide here: https://github.com/FortAwesome/Font-Awesome/issues/18775#issuecomment-1073217558

**Comment by ghost** [comment]
> We are trying to sort this out without breaking backward compatibility but it is a little bit trickier
>
> There is also a workaround, but we don't know yet if it will fix 100% of the cases and requires effort on developer side. If you want to test, I've provided a step-by-step guide here: https://github.com/FortAwesome/Font-Awesome/issues/18775#issuecomment-1073217558

I'm not in a hurry, I'll wait for this issue to be fixed by Font Awesome. The issue happens quite rarely, and you just have to reload the web page again to temporarily fix it.

> As per dart sass, I can see that this has been partially fixed in 1.38.0, please make sure that you are using >= 1.38.0

I'm using the `1.49.9` version  of [sass/dart-sass](https://github.com/sass/dart-sass). So, I guess their partial fix doesn't really work for me!

```json
{
	"devDependencies": {
		"sass": "^1.49.9"
	}
}
```

>

**Comment by danon** [comment]
I also have `1.49.0` and I'm experiencing the bug too.

@BrokenSourceCode Have you found a way to fix this bug on our end?

**Comment by ghost** [comment]
> I also have `1.49.0` and I'm experiencing the bug too.
> 
> @BrokenSourceCode Have you found a way to fix this bug on our end?

Not really, but you can try to programmatically add the rule `@charset "utf-8";` at the **VERY TOP** of your CSS output files (**AFTER** the processing of [dart-sass](https://github.com/sass/dart-sass)). You may use something like [gulp-header](https://github.com/gulp-community/gulp-header) or [webpack](https://github.com/webpack/webpack) (with the [`BannerPlugin`](https://webpack.js.org/plugins/banner-plugin/)).

By the way, don't forget to put the HTML tag below inside the `<head>` tag:
```html
<meta charset="UTF-8">
```

**Comment by tagliala** [comment]
Hi @BrokenSourceCode ,

as per my understanding, `@charset "utf-8";` **and** `<meta charset="UTF-8">` are not enough to fix 100% of the cases. Ref: #18775

@GameCharmer is testing a fix that we can implement without breaking backward compatibility (on expense of compile time and scss source size), so I would like to ask if there are any news (but I'm assuming no news is good news)

**Comment by GameCharmer** [comment]
@tagliala So far so good! We haven't received a single message about it from our users, and none of the development team has ran into it since implementing your fix.  You definitely nailed it! :D

**Comment by OwenPattison** [comment]
@GameCharmer @tagliala This sounds great thanks for the continued effort.

Is there a timeline for this release?

**Comment by robmadole** [comment]
No ETA yet. @tagliala has done a lot of the ground work here but we'll need to make sure we've got this nailed down before we release it.

**Comment by tagliala** [comment]
> we'll need to make sure we've got this nailed down before we release it.

Agree. An application I maintain suffers of the same issue, but I would definitely avoid such a change if it is not going to fix  100% of the cases.

More feedback would be welcome.

Here it is a guide on how to test the workaround manually:

> I've tried to provide a step by step fix here: https://github.com/diowa/ruby3-rails6-bootstrap-heroku/pull/540
> 
> This is based on Font Awesome Free, so steps for Pro should be a little bit different
> 
> At the end of the procedure, double check that variables in the output css have the proper escaped entities:
>  - ✅ `.icon::before { content: "\f26e"; }`
>  - ❌ `.icon::before { content: ""; }`
> 
> NOTE: This change will make harder to update Font Awesome

Feel free to ask any question

**Comment by Iridium-again** [comment]
Storefront theme is mainly working really well. But all 3 handheld footer bar icons disappear in Safari only.
When viewed on an iPhone, the Handheld footer bar icons are all visible in Opera, Vivaldi, Firefox, Chrome.
But in Safari they turn into tiny blank rectangles, with no indication to site visitors as to what they are for, unless the visitor happens to tap on one of them, then a little explanatory message pops up.
Wondering if it's my fonts plugin causing this
![Craftsteading – Use plants to make anything](https://user-images.githubusercontent.com/109767078/180318034-7c63436e-f7ce-40db-848c-d13a3597de63.png)

**Comment by viniciusdeliz** [comment]
I can confirm that this happening with sass versions `1.50.1` and `1.54.0`. With valuable information provided by @tagliala, I've also tested and confirmed that the issue doesn't happen with sass `1.10.3`. 

I have tried inserting `@charset "UTF-8"` with webpack BannerPlugin at the beginning of the compiled CSS but that also haven't worked (meta charset was already in-place in the HTML). Apparently the font-awesome already carries values such as `fa-ambulance:before{content:""}` at the moment that the charset rule is applied in the stylesheet. Also tested a bunch of other webpack plugins and compilers/loaders but I couldn't figure out a solution.

Weirdly enough, this appears to happen only when the project runs in production. In staging, font-awesome:before content appears with the "\f00" expected value. I individually tested plugins and settings and couldn't trace why this happens.

**Comment by nirt04** [comment]
Hey guys, we don't use font-awesome but we came across your issue while trying to solve this problem with our own library
so how you realized sass is decoding the Unicodes in build time and causing broken content to the classes

we work around this by importing the CSS icons files directly to js and skipping the sass-loader.

before:
<img width="215" alt="image" src="https://user-images.githubusercontent.com/47310131/185902047-c6aa0c3e-174b-411d-8806-67a79d00145a.png">
after: 
<img width="368" alt="image" src="https://user-images.githubusercontent.com/47310131/185902127-d3c5b93c-8b28-4e02-b33e-03ed22234cc8.png">
<img width="467" alt="image" src="https://user-images.githubusercontent.com/47310131/185902458-c3abaaea-8603-462f-ae50-90577cac5910.png">

**Comment by GameCharmer** [comment]
@tagliala I updated from 6.0.0 to 6.2.0 with no luck using the previous fix. Swapping the function results in no icons at all (not even the broken squares), and updating the variables results in the F132 string being displayed as plain text.  Any ideas?  I'm still running the same SCSS compiler.

**Comment by tagliala** [comment]
> I'm still running the same SCSS compiler.

which is scssphp?

Please let's continue on #18775 and please retry https://github.com/FortAwesome/Font-Awesome/issues/18775#issuecomment-1070853882

I need an updated version of https://github.com/FortAwesome/Font-Awesome/issues/18775#issuecomment-1070882454

**Comment by GameCharmer** [comment]
scssphp/scssphp v1.10.1

18775 is still locked for me.

with option false
```
@charset "UTF-8";
.icon::before {
  content: "";
}
```

with option true
```
.icon::before {
  content: "\f26e";
}
```

Update: Updated to v1.11.0, identical output as above

**Comment by tagliala** [comment]
@GameCharmer the output looks ok and I do not see any possible issue with that specific compiler

I cannot reproduce the issue, my previous use case works good with 6.2.0 and the output CSS is fine

Steps to manually fix are described in the ~comment~ individual commit messages. You should redo the work with 6.2.0 files, otherwise new icons will not be available

https://github.com/diowa/ruby3-rails6-bootstrap-heroku/pull/540

### Screen

<img width="671" alt="image" src="https://user-images.githubusercontent.com/556268/192590234-c99b395f-0b52-4427-a6ff-3134e03d922d.png">

### Output CSS (dart sass 1.55.0)

```css
.fa-github:before {
  content: "\f09b";
}
```

As far as your output is like the one above, then the cause of this new issue may be different

For any other attempt on my side, I need a remotely reproducible test case, or a minimal docker container, since I do not have the libraries needed to work with scssphp

**Comment by viniciusdeliz** [comment]
> I also have `1.49.0` and I'm experiencing the bug too.
> 
> @BrokenSourceCode Have you found a way to fix this bug on our end?

Any luck with the 1.54.0 update in sass? The change appears to be the same of the 1.39.0 that was supposed to fix this, but maybe it works this time? https://github.com/sass/dart-sass/blob/main/CHANGELOG.md#1540

**Comment by tagliala** [comment]
> Any luck with the 1.54.0 update in sass? 

There is a partial fix in 1.38.0, as per https://github.com/FortAwesome/Font-Awesome/issues/17644#issuecomment-1071144856

I can confirm that in 1.55.0 this issue has been partially fixed, in the meaning that codes in the private area are correctly escaped, letters and numbers aren't (would this be a problem? 🤷🏻‍♂️), and I can't see unescaped unicode characters in the output CSS, so the `@charset` rule is not being emitted

```
$ npx sass --version
1.55.0 compiled with dart2js 2.18.1
$ npx sass node_modules/@fortawesome/fontawesome-free/scss/fontawesome.scss
```
Partial output:
```css
.fa-at::before {
  content: "\@";
}
/* ... */
.fa-0::before {
  content: "\30 ";
}
/* ... */
.fa-p::before {
  content: "P";
}
/* ... */
.fa-snowflake::before {
  content: "\f2dc";
}
```

Anyway, adding the `@charset` rule does not always work, in the meaning that *sometimes* you can still experience this issue (ref #18775), and, in addition to that, some upstream minifiers on top of Sass may decide to strip the rule out (webpack-based bundlers may do that), so probably a fix on Font Awesome is still needed for this issue

**Comment by ghost** [comment]
> > I also have `1.49.0` and I'm experiencing the bug too.
> > @BrokenSourceCode Have you found a way to fix this bug on our end?
> 
> Any luck with the 1.54.0 update in sass? The change appears to be the same of the 1.39.0 that was supposed to fix this, but maybe it works this time? https://github.com/sass/dart-sass/blob/main/CHANGELOG.md#1540

@viniciusdeliz I updated sass to version 1.54.0 on July 30th (currently my version is 1.55.0), and since then I've already encountered the weird characters on refresh, several times, randomly. So, on my side, the issue is still not fixed, but it seems that the problem occurs less often than before (maybe it's just an impression).

> Anyway, adding the `@charset` rule does not always work, in the meaning that _sometimes_ you can still experience this issue (ref #18775), and, in addition to that, some upstream minifiers on top of Sass may decide to strip the rule out (webpack-based bundlers may do that), so probably a fix on Font Awesome is still needed for this issue

@tagliala I agree with you: a Font Awesome's fix is still needed for this issue. 👍

**Comment by domenkozar** [comment]
I've hit this one using parcel: https://github.com/parcel-bundler/parcel/issues/8713

**Comment by domenkozar** [comment]
I noticed that the css file doesn't get interpreted as css, so I investigated deeper into it. Opening up the debugger tools reproduced it consistently, so that was the first clue.

I had a theory that `content-type: text/css; charset=utf-8` should be set as the header, which wasn't the case when I was serving files from nginx.

Following `nginx.conf` in the http section helped:

```
            source_charset utf-8;
            charset utf-8;
            charset_types text/html text/xml text/css text/plain text/vnd.wap.wml application/javascript application/rss+xml;
```

Now https://cachix.org will finally not get garbled fonts :rocket:

**Comment by florent-cdl** [comment]
Found that post that was exactly my problem : encoding mixin error with libsass
https://github.com/sass/libsass-net/issues/24
The workaround work great
@mixin pseudoFontIcon($content) {
    content: #{"\""+$content+"\""};
}
a:before {
  @include pseudoFontIcon(\e003); // without single/double quotes here
}

**Comment by magynhard** [comment]
For others, facing the same issue:

The problem with displaying the icons incorrectly when contained in unicode characters is Chrome related and happens only in Chrome and other Chromium based browsers (e.g. Edge or Vivaldi). I could not reproduce it on Firefox or Safari.

It is not related to FontAwesome, it is related to all Iconfonts using unicode. (Im my case my own font)

It can even be reproduced regularly the following way:
* Open your browser
* Open your website
* Open Chrome Dev Tools (F12)
* Reload Page (only F5 or Reload button, not Force Reload)
* Icons are displayed incorrectly

It works only once, if you want to repeat, you need to close the developer tools, reload the page, then reopen the developer tools again and reload again.

It seems, that chrome does not check or detect the format of the CSS correctly when the page is reloaded with developer tools open at the first time.

For me, the best solution was - as mentioned by others before - to add
```
@charset "UTF-8";
```
at the beginning(!) of the .css file to workaround the Chromium engine bug.

**Comment by tagliala** [comment]
Just to mention (again) that adding `@charset "UTF-8";` will not fix 100% of the cases

I suggest to consider also changing response header via nginx or whatever proxy server: https://github.com/FortAwesome/Font-Awesome/issues/17644#issuecomment-1362785107

**Comment by magynhard** [comment]
If using escaped utf8 characters beginning with `\f` adding the charset to the css header seems to work always, but maybe not when having the real utf8 characters inside the css. I think changing the web server header to fix this is no good solution. In my opinion you should convert to escaped characters and add the charset. Especially, when the developer has no option or no easy way to configure the webserver. So you are independent of whatever webserver is used and have the control in your web project instead on the webserver layer.

**Comment by tagliala** [comment]
> In my opinion you should convert to escaped characters and add the charset

I'm 100% for this solution, and I think that FA should output escaped chars, but unfortunately it's a breaking change :(

Upgrading to dart sass > 1.55 should also help (or maybe even fix)

**Comment by nsunga** [comment]
> ### Solution
> if you have PUA codes (e.g., `content: ""` instead of `content: "\f26e"`) in the output CSS, check that:
> 
> 1. The served CSS file starts with `@charset "UTF-8";`; **AND**
> 2. The HTML file referring the CSS has the `<meta charset="utf-8">` meta tag
> 
> [#14660 (comment)](https://github.com/FortAwesome/Font-Awesome/issues/14660#issuecomment-790475966)
> 
> # ⚠️ update
> Unfortunately this is not enough. It may fix 99% of the cases, but not 100%

hello @tagliala !

so sorry to bring up such an old comment

I did read the update, but with this "fix" does that mean the minified code results in the unicode string (ie. "\f26e")?

at the moment, the minified css I have are all PUA (Private Use Area) codes even with the "fix"

**Comment by tagliala** [comment]
> I did read the update, but with this "fix" does that mean the minified code results in the unicode string (ie. "\f26e")?

No, if you want a fix like that, it is a little bit more complicated at the moment and it will require a new fix after each update, please check the comments here: https://github.com/diowa/ruby3-rails6-bootstrap-heroku/pull/540

**Comment by lindalinya** [comment]
> Hey guys, we don't use font-awesome but we came across your issue while trying to solve this problem with our own library so how you realized sass is decoding the Unicodes in build time and causing broken content to the classes
> 
> we work around this by importing the CSS icons files directly to js and skipping the sass-loader.
> 
> before: <img alt="image" width="215" src="https://user-images.githubusercontent.com/47310131/185902047-c6aa0c3e-174b-411d-8806-67a79d00145a.png"> after: <img alt="image" width="368" src="https://user-images.githubusercontent.com/47310131/185902127-d3c5b93c-8b28-4e02-b33e-03ed22234cc8.png"> <img alt="image" width="467" src="https://user-images.githubusercontent.com/47310131/185902458-c3abaaea-8603-462f-ae50-90577cac5910.png">

Actually, this case is not from sass-loader but dart-sass. I have looked up many, many issues about this case, which are all over the major open-source warehouses such as sass, sass-loader, and so on.

The source of the problem is not the sass-loader, the sass-loader just calls the sass command, so the source of the problem is the default encoding of dart-sass (you can get this result just with sass-loader and file-loader), which stubbornly encodes characters that do not need to be encoded. So "\xxx" in our code is translated into a garbled character after dart-sass compilation.

Unfortunately, dart-sass could have had an elegant solution to this problem, however, their developers were so stubborn that they believed their default coding could not be changed. They think this effect is so great that they act like bandits and demand that person who asked the question provide stable replication steps for an occasional bug.

Guess the ultimate source of this problem can actually be attributed to the browser, the browser for the CSS encoding guess error, resulting in even if you have added html encoding and CSS charset, the font icon parsing still occurs problems.

The final solution, as this man says, is to **Change your sass file to a CSS file**. **Extract all CSS styles that contain Unicode characters into a CSS file, not a Sass file. The CSS file does not need to go through the sass-loader, so the stubborn dart-sass cannot corrupt it**

**Comment by nsunga** [comment]
Update on my end:

I ended up using a postcss plugin - https://github.com/lcoronelp/postcss-sass-unicode

I will say that its not great, though **because this is what looks like is happening**:
1. dart-sass is taking the unicode characters and encoding them to the square things []
2. then, the postcss plugin takes the square things and then decodes them back to the unicode characters in the final css

So now when the browser loads in the css, it doesnt have to do its decoding for whatever the square [] is - its just reading the value.

Until theres a popular solution, ive just been using the above plugin for now

**Comment by lindalinya** [comment]
> Update on my end:
> 
> I ended up using a postcss plugin - https://github.com/lcoronelp/postcss-sass-unicode
> 
> I will say that its not great, though **because this is what looks like is happening**:
> 
> 1. dart-sass is taking the unicode characters and encoding them to the square things []
> 2. then, the postcss plugin takes the square things and then decodes them back to the unicode characters in the final css
> 
> So now when the browser loads in the css, it doesnt have to do its decoding for whatever the square [] is - its just reading the value.
> 
> Until theres a popular solution, ive just been using the above plugin for now

Maybe you can try my idea if your font icon related styles are all in a sass file. Just change this sass file to a CSS file to make it escape dart-sass

**Comment by damirbogdanov** [comment]
> Update on my end:
> 
> I ended up using a postcss plugin - https://github.com/lcoronelp/postcss-sass-unicode
> 
> I will say that its not great, though **because this is what looks like is happening**:
> 
> 1. dart-sass is taking the unicode characters and encoding them to the square things []
> 2. then, the postcss plugin takes the square things and then decodes them back to the unicode characters in the final css
> 
> So now when the browser loads in the css, it doesnt have to do its decoding for whatever the square [] is - its just reading the value.
> 
> Until theres a popular solution, ive just been using the above plugin for now

thank you @nsunga, this approach seemed to help in our case and helped to resolve an issue

**Comment by vineeth-ktp** [comment]
Issue is often(50% of the time) occurring with Mac Chrome. I have tried adding charset, but it still fails. The icons gets added by CMS and we don't have a way to determine which of them to preload. We are using sass and not with dart as it is not supported by CMS. Is there any other solution suitable here?

**Comment by tagliala** [comment]
> We are using sass and not with dart as it is not supported by CMS

This is supposed to happen with self-hosted css files compiled from Sass, so it should mean that user have the ability to make changes to the toolchain used to compile the css output.

You can:
- Downgrade to an older version of Sass not affected by the issue, ref: https://github.com/FortAwesome/Font-Awesome/issues/17644#issuecomment-796114103
- Fix the provided css by transforming some values. Steps at https://github.com/diowa/ruby3-rails6-bootstrap-heroku/pull/540

**Comment by viniciusdeliz** [comment]
For fellows using webpack and facing this issue, I've realized that while on localhost my stylesheet always had the icons in place. So I found that in the stylesheet of my built application I had the following output:

![webpack-fa](https://github.com/FortAwesome/Font-Awesome/assets/5748616/cfd6cac5-6478-48b7-bf1a-1f07078398b0)

So testing a bit of options, I figured that the culprit of my setup was the sass-loader set for _.scss_ files. So on my **webpack.config.js** for my `/\.scss$/` rules, I included the following `sassOptions` entry in the options object for the sass-loader:

```
{
    loader: 'sass-loader',
    options: {
        sassOptions: {
            outputStyle: 'expanded',
        },
    },
},
```

Now the stylesheet output for my build is the following:

![webpack-fa-sol (1)](https://github.com/FortAwesome/Font-Awesome/assets/5748616/c71269c9-abba-47be-af86-cd4732367236)

Hopefully this can help someone!
